### ISSUE TITLE
Create an additional shaded API jar

### ISSUE THREAD

**Issue by kashike** [SpongePowered/SpongeAPI#679]
It would be nice to have an API JAR with dependencies shaded for developers who do not wish to use a build system.

**Comment by Mumfrey** [comment]
Wouldn't that encourage people to not use a build system though?

**Comment by ElgarL** [comment]
Some of us old-timers do not like build systems ie  maven/gradle

**Comment by Mumfrey** [comment]
I'm an "old timer" so to speak, and still use ant and ivy for things in my own projects. Why would you *not* use a build system, it makes no sense?

**Comment by Zidane** [comment]
While @progwml6 will argue against this wholeheartedly (and I'm sure many others will), I see the points being made.

Build systems are wonderful and I'd never recommend anyone to NOT use one but its just that, a recommendation. It shouldn't be a requirement for SpongeAPI to simply HAVE to have a build system (or go out and fetch the libs yourself).

While the shaded API jar won't be the norm nor be used in the impl, there is nothing preventing us from providing a -all jar for the API.

It comes down to choice, should we punish people who use their freedom in software choice by choosing to NOT use a build system by having to seek out all the libs themselves?

**Comment by Mumfrey** [comment]
It's not punishing people to simply observe best practices. If people want to deviate from best practice that's entirely their own right, but a platform should encourage good practices and not using any form of build system doesn't fit in that space.

**Comment by ElgarL** [comment]
Oddly I HATE 'best practices' and 'norms of programming'. Learn them of course but the greatest inventions and leaps forward were made by people not trained in the field. People who were not constrained by the limitations of being taught what is right and what is wrong.

Learn how to do things correctly, then go out and break everything by doing it your way.

**Comment by progwml6** [comment]
Use a build system. We agreed from the beginning of the project that there will not be a shaded API jar

**Comment by Zidane** [comment]
@Mumfrey 

We do encourage and have done so this entire time. You'll not see one ounce of non-build system usage throughout our docs and elsewhere.

Again, besides principle (which one could argue) is there simply any reason why there couldn't be a shaded API jar provided for download?

**Comment by Zidane** [comment]
@progwml6 

When an issue is being discussed, that doesn't mean you can close it.

**Comment by Mumfrey** [comment]
> Oddly I HATE 'best practices' and 'norms of programming'. Learn them of course but the greatest inventions and leaps forward were made by people not trained in the field. People who were not constrained by the limitations of being taught what is right and what is wrong.
> 
> Learn how to do things correctly, then go out and break everything by doing it your way.

That's a great ethos and I respect it, but it still doesn't really explain why you would not use the tools provided. If you're not using a build system and doing everything by hand, then it's not fair to ask for special provision be made the existence which could mislead others into believing it's a good idea.

I mean, I can understand you not liking maven or gradle, I don't either. But using (as a minimum) ant in your project just saves you from doing the entire build process by hand every time, and will do things like fetch all the dependencies for you.

**Comment by ElgarL** [comment]
Mumfrey I do use Ant but I believe an API should provide all of its dependencies to the end user (programmer). It should never be returning unknown objects to an IDE without also providing the corresponding class.
An API should be complete in that it can be added as a dependency and fully used without any further steps required by the programmer. Depending on a build system breaks specific features in IDE's like jar exporting.

Build systems are great, but should never be required.

**Comment by Mumfrey** [comment]
If you're using ant then why aren't you fetching dependencies with ivy? It'll pull everyhing you need down in one swoop and you have full control of where those jars go (ivy "retrieve" can be configured to just dump everything in a folder called "libs" for example (which is what I do)). Then you're free to do what you want with it.

**Comment by Mumfrey** [comment]
By the way, if you're not familiar with it, [here's one of my own ant scripts](https://github.com/Mumfrey/LiteLoaderInstaller/blob/master/ant/build_installer.xml) which shows how to pull down the deps with ivy.

The key parts are the [retirieve step](https://github.com/Mumfrey/LiteLoaderInstaller/blob/master/ant/build_installer.xml#L31) (which dumps all the requisite jars in "libs") and [this funky shorthand](https://github.com/Mumfrey/LiteLoaderInstaller/blob/master/ant/build_installer.xml#L24-L28) which then adds all the fetched libs to your build-time classpath.

The actual dependencies are just specified in a [simple xml file](https://github.com/Mumfrey/LiteLoaderInstaller/blob/master/ant/ivy.xml) and transitive dependencies are handled automatically.

**Comment by ElgarL** [comment]
Because I currently have no dependencies as I'm testing out features of Sponge. I should not be required to use a build system or script to simply export a jar.
As to Ivy, is that something new? Another download/install? More levels of complexity which are not needed. May be handy but not needed.

**Comment by Mumfrey** [comment]
Nope. Ivy's older than fire, it's kind of the companion product to ant and you should check it out. As I said, I'm a miserable old so-and-so and basically stick to ant+ivy for all my own stuff. I certainly wouldn't call it a layer of complexity, it's kind of the opposite, basically it *removes* complexity because you just say "download all this crap" and lo, crap is downloaded.

**Comment by ElgarL** [comment]
I just looked it up and really I have no need for it. I wrote an Ant script years ago to handle downloading deps and I've never had to alter since.
I can download Towny onto a clean install of Eclipse with zero other requirements. Build a fully working jar without any need for anything else. Its how it should be. Simple and clutter free.

**Comment by Mumfrey** [comment]
That seems like cutting off your nose to spite your face. Things can be extremely simple, if you already acknowledge that you need to download deps automatically, then using a dependency manager *provides* that simplicity. Are you downloading everything using raw URLs or something? That seems to be *more* complex than adding a single jar to your project then just declaring the dependencies in a one place.

You've kind of lost me here because on the one hand you seem to want to keep things simple, but seem to be doing so in the most complicated way possible.

**Comment by Mumfrey** [comment]
I'll happily write your ivy configuration for you if you want, can you link your project's source or is it private?

**Comment by ElgarL** [comment]
I would rather write a simple one line in an Ant script to download a dep than install another app. Not sure if Ivy comes packed with Ant these days, but it looks like it didn't.

Also I download deps from my own server. To rely on anyone else hosting is suicide.

**Comment by Mumfrey** [comment]
It's an ant plugin, so you just bundle the jar with your project, then import it in your ant script.

Would you at least consider seeing the other side here? It seems like a lot of your pain is effectively self-inflicted because you would rather do things a difficult way than a simple one. Whilst I completely agree that forcing someone to use maven or gradle is not something we should do (let's face it, people using scala will want to use sbt for example), managing dependencies by hand is not something anyone wants to do. There are off-the-shelf solutions to deal with what you want.

Ivy is a simple, repository-agnostic dependency manager that is really easy to set up and will remove all your pain going forward, and as I've said I'll happily set it up for you now so that there's zero effort on your part to make the transition (in fact the ulterior motive is that it'd give people a good go-to example for others who don't want to use maven or gradle).

Ivy supports everything from full blown maven repos to simple "get lib from this url" declarations, and effectively all you'd do (assuming you have your deps declared in your ant file, with urls to same) would be to move that list into a separate file (also xml).

I think everyone on this team could engage with your point if there were a cogent argument behind it, but at the moment I don't see how we can come around to your point of view when there doesn't seem to be any benefit to your approach, and you are in fact making things a more complicated than they really need to be.

**Comment by ElgarL** [comment]
My point is, I don't need anything extra to handle it. I have Ant built into Eclipse and it does everything I need. I could install Ivy to do that then I could install something else to do whatever and before long you have 100 different little programs installed, each one doing one little thing.

I have ONE install at the moment. I run Eclipse and I'm done. Nothing else. No other plugins and nothing else to add complexity or to go wrong.

These days everyone is after the latest this or that. Got to have the most fashionable build tool or plugin to auto something or other. In the field of programming the old adage of KISS still applies and should ALWAYS apply.

An API when added to a build path should not be exposing unknown/undefined objects to an IDE.

**Comment by jk-5** [comment]
You're a bit overreacting now. There is no 100 little programs.

It's just the ivy jar for ant which will download all the files for you that you download manually from ant in the towny ant script. If you still don't want to touch ant you can always download all the dependencies by hand or using ant.

**Comment by ElgarL** [comment]
Its called an exaggeration to explain a point.

At one point it was only Ant you needed to install. Oh and the Ant plugin for Eclipse. Now its Ivy, or its Maven and the Maven plugin and whatever else.

I don't need nor want any of it. A decent IDE and an API is all I should need.

**Comment by kenzierocks** [comment]
> An API when added to a build path should not be exposing unknown/undefined objects to an IDE.

What do you mean by that? You mean things that don't have the javadocs you'd get by using ivy?

**Comment by Zidane** [comment]
@kenzierocks 

He means the fact that using just the API jar will not work as Optional is from guava which is a class not in the API jar. A build system obviously rectifies this as it has guava auto-download and the end user is none-the-wiser.

**Comment by Mumfrey** [comment]
I have to agree with jk-5.

> My point is, I don't need anything extra to handle it. I have Ant built into Eclipse and it does everything I need. I could install Ivy to do that then I could install something else to do whatever and before long you have 100 different little programs installed, each one doing one little thing.

As I said, you don't "install" ivy, because it's an ant plugin you just bundle it with your project, then in your ant script you import the jar. Nothing extra needs installing at all.

> I have ONE install at the moment. I run Eclipse and I'm done. Nothing else. No other plugins and nothing else to add complexity or to go wrong.

This would still be the case, the only change would be you could also install IvyDE (an eclipse plugin) which would read the classpath out of the ivy file automagically. However that's *optional* and just having the ant script use ivy isn't another step at all, you'd still just run the ant file like before.


> These days everyone is after the latest this or that. Got to have the most fashionable build tool or plugin to auto something or other. In the field of programming the old adage of KISS still applies and should ALWAYS apply.

I completely agree, but this is where your argument doesn't make sense. KISS says use the simplest solution, but your solution is more complex than need be.

> An API when added to a build path should not be exposing unknown/undefined objects to an IDE.

How is downloading the jars automatically any different than bundling extra classes opaquely within the API jar?

**Comment by ElgarL** [comment]
Because with a shaded API jar there are no downloads. You are not downloading anything because the API is providing everything it should be. There are no additional deps, nothing extra to install or configure
Works straight out of the box on a vanilla IDA install.

**Comment by Mumfrey** [comment]
Indeed, so does my proposed solution. Your ant script is (as you mentioned) above, already grabbing everything you need, so this would just automate that and grab the transitive dependencies as well.

**Comment by ElgarL** [comment]
As I already said, I'm not using any script at the moment.

Do as you please but I'm not porting/coding for Sponge as it currently stands.

**Comment by Mumfrey** [comment]
Sorry I'm really confused. Are you using ant to download the dependencies for your project or not?
You mentioned earlier

> However I've always managed my own download/management of deps with Ant.

And I'm not proposing to change that.

> Do as you please but I'm not porting/coding for Sponge as it currently stands.

I don't see why you would write off an entire API based on wanting to stick to something which is only going to become harder to maintain going forward. And this isn't about "as we please". I'm honestly trying to help you here and I don't see why you are so combative about the whole thing. Feel free to drop onto IRC later and we can chat about it.

As I said above, I think we could all get on board with your idea if it made any sense, which given that there's a simpler solution literally sitting in front of you (did I mention already that I'll do this for you so you don't need to lift a finger, and it doesn't change any of your workflow) it just doesn't seem to.

I'm sorry you feel that way but if you don't want to accept help when it's offered openly then I'm genuinely baffled.

**Comment by Tzky** [comment]
Mumfrey, if i got him right this is *not* about the way dependencies are added but about the fact that additional dependencies are needed. He wants only one dependency (spongeAPI), as he thinks an API should be the only dependency for plugins.

**Comment by Mumfrey** [comment]
I understand what he wants, I'm just trying to get to the root of why he wants it. Since it represents a fairly hefty deviation from best practice (and, let's face it, sensible practice) there needs some solid reasoning for undertaking it.

Here's an analogy:

*Let's assume you're a car manufacturer. You make and sell cars in a standard format, with standard controls that people understand. The standardisation is what makes the car safe and usable to people who know how to drive cars in the normal way.*

*Someone comes to you, and says. "I've taught myself to drive by laying in the drivers seat back-to-front, and looking where I'm going using mirrors. Can you mount mirrors in the back of the seat so that I don't have to add my own every time.*

Of course, the first question you ask isn't *"what shape should the mirrors be?"*, you ask *why don't you just try facing forwards. The user points out that they don't like using cruise control and ABS, but you point out that you don't need to use cruise control or ABS to operate the vehicle, you just need to sit facing the right way and can ignore those features which make driving slightly more streamlined if you want a more "hands-on" driving experience. Just sitting facing forwards is the simplest option, in fact it's simpler than your current solution because you don't have to fit your own mirrors every time you buy a car, and can in fact see more because you can see out of the windows as well.

The user protests that they like their mirrors, and could easily gain the "seeing out of the window" functionality by adding more mirrors, but you point out that if they just face the other way then they get this all for free, and moreover will have a nicer driving experience in general in the future.

The manufacturer hasn't actually said "no" yet, but trying to convince the user to adopt a simpler, easier solution should be the path of first resort, because really you can see that the user is actually making things harder for himself unnecessarily, and the features he proposes to suit his use-case may harm someone else (maybe people will get mirror glass in their back, or just be plain confused by the mirrors in the seat and think that *they too* need to drive facing backwards.

Of course, if nothing comes of the discussion, and the user is determined to preserve their somewhat eclectic and challenging mode of driving, then eventually the manufacturer might relent and make the mirrors-in-seat an built-in optional extra on some models. But the duty of care is to scope out the simple solutions first, the one which provides the greatest benefit to the end user.

It's also probably worth waiting for @kashike, who opened the issue to weigh in with his/her opinion too, since the discussion has thus far been dominated by @ElgarL and I discussing Ivy, and we haven't actually gotten to the original use-case of the OP.

I personally dislike being "forced" to use any one solution as well, so I can completely see where @ElgarL is coming from, but as I mentioned above, there are fundamental contradictions in his argument which bear further examination before going any further. Indeed, much as I hate to admit it, Maven is actually probably the best choice for his specified goals (just install the IDE, no extra steps) because it's built in to absolutely everything and it's stupidly easy to use. The only drawback with maven is that its repository architecture isn't pluggable so he'd have to muck around with it if he wanted to stick with the "everything on my server" setup. Ivy on the other hand is super configurable and could easily be a drop-in solution for his use case with pretty much zero effort.

We'll wait and see what other opinions surface anyway.

**Comment by boformer** [comment]
For simple plugins, a build system is a total overkill. Especially if you are new to Gradle/Maven.

I agree that when you publish the source code of a plugin, a build script is a must-have.

I don't see the benefit of a build system for little (private) experiments. For example, I have a "SpongeTest" plugin in my IDE to test the example code I'm writing for the SpongeDocs.

I was working on a wizard to simplify the setup of the gradle script, similar to the [LibGDX setup](https://github.com/libgdx/libgdx/tree/master/extensions/gdx-setup) (they also require gradle).

!(Wizard)[https://forums.spongepowered.org/uploads/default/_optimized/fa6/d0e/b21ba15389_690x388.png]

**Comment by Mumfrey** [comment]
> For simple plugins, a build system is a total overkill. Especially if you are new to Gradle/Maven.

If you read the backlog, this point is not in contention. However, using a dependency manager like Ivy is super simple and doesn't have any of the issues which a shaded jar would have.

As for your other points, please keep general unrelated discussion out of the issue thread, otherwise it's likely to wander off topic. This isn't about whether using a build system is good or bad, it's about whether a shaded jar is needed to support those not using a build system (which it isn't).

**Comment by boformer** [comment]
@Mumfrey Ivy is just another tool that stops people from starting plugin development. I think a shaded API jar simplifies the process.

**Comment by Mumfrey** [comment]
Again, that's off topic though, because we're never going to actively encourage people to start developing without using a build system, it's just silly. It'd be like encouraging people to start writing a plugin without an IDE. The "it's just simpler" use-case is and always was off the table. The active point of discussion is people with existing workflows and the simplest way for them to integrate SpongeAPI into that.

In this instance, the active platform is ant, and plugging Ivy into ant is a no-brainer.

**Comment by ElgarL** [comment]
@boformer is most definitely ON topic.
@Mumfrey you keep talking about it being a simple case of 'adding ivy' or whatever else you define as simple. Adding anything is the WHOLE point. It should not be required to add anything to write a plugin for sponge. Sponge should be the complete API and all you need.
Requiring developers to add anything just to use the API is making it less appealing and harder to use.

KISS (Keep It Simple, Stupid), and about the simplest you can get is adding Sponge to your build path.

**Comment by Mumfrey** [comment]
> Adding anything is the WHOLE point. It should not be required to add anything to write a plugin for sponge. Sponge should be the complete API and all you need.

But this is the point which I dispute. *Why* should it be that way, it doesn't make any sense to me. I kind of get where you're coming from but still don't see the use-case. New users can use the SDK so it doesn't benefit them, so that only leaves people with existing build chains and I already pointed out that there's a better solution for that in your particular case.

> Requiring developers to add anything just to use the API is making it less appealing and harder to use.

But appealing to whom? New users aren't going to be put off by the SDK, and since you're using ant already there's a straightforward answer for you too. I guess what I'm asking is, who is this **for**? I still don't get it.

**Comment by ElgarL** [comment]
I guess you never will then.

**Comment by Mumfrey** [comment]
I'm trying not be closed-mined about this, but you haven't made an argument yet which makes it clear why this is necessary.

**Comment by ElgarL** [comment]
I've said all I can. Some can see the reasoning some can't, but I have no other way to explain it.

**Comment by Mumfrey** [comment]
I must apologise then, I'd like to help but I really can't see how.

**Comment by Mumfrey** [comment]
Would either of these options work for you: 

* We add a task in the API `build.gradle` file to emit a fully shaded jar, which would allow you to build it and host it on your server, which is what I assume you want to do ultimately.
* We do that, but I set up my jenkins (ci.liteloader.com) to also build that shaded jar and give you a place to download it from.

If either of those are helpful (or both), then I can set that up. It keeps the fully shaded jar off the official Sponge repositories but gives you somewhere to get what you need.

**Comment by ElgarL** [comment]
Its a nice offer and I believe a shaded jar is the correct way to go, but I feel it needs to be an officially available option. I think it should be the default option over a build system favoured one.

**Comment by Mumfrey** [comment]
I can't see us encouraging people to avoid best practices though. If you want the artefact available for your own purposes then then above offers stand and you can just poke me on IRC (I'm online when the M in my name is uppercase).

For ordinary dev we have the SDK and that will basically always be the recommended option. You may wish to await confirmation from a project leader but I'm almost certain they will say the same thing, sorry.

**Comment by ElgarL** [comment]
I've never seen a 'best practice' defined to require a build tool.

**Comment by ConnorStone** [comment]
Don't been to bud in, but if I can help, I definitely see the point on both sides. I believe what @ElgarL is trying to say in that SpongeAPI should be as simple to setup as bukkit, if the programmer WANTS it to be.

For example, if you wanted it simple: 
* API jar to the build path
* Profit

Where as if you were managing a complex set of multilevel dependency projects, such as Sponge and some of the bigger plugins, you might opt for a much more complex build system with control over aspect of the build and all the dependencies.

I see @Mumfrey 's side of this, we don't necessarily want to discourage learning to use build scripts, but either way I believe anyone, novice or "old timer" should have a choice, even if it means they are being very simple minded about a specific project case scenario. I don't believe this is about Towny, but that it should be simpler for people if they want it to be.

We should remember for people who have never heard of Maven, Gradle and Ant, it all a bunch of mumbo jumbo and they would rather start of simpler. I remember I first learned java cause I wanted a bukkit plugin, and the tutorials were so easy to follow. For new coders, a build system would be too much.

**Comment by ryantheleach** [comment]
Would bundling the dependencies in a shaded jar hurt anyone that doesn't want to use it?

Does it hurt the sponge project at all (apart from having to serve the extra bandwidth required)?

If not I don't see why it shouldn't be an option. Just a discouraged one.

Also btw @ElgarL not having a maven repo for towny is an utter pain in the ass for anyone that was developing bukkit plugins that tried to interact with towny. Instead of just specifying a repo location, a name, and a version people would need to hunt down the url and either download it or make special provisions in order to download towny to fulfil their towny dependencies.

Whilst I dislike having to chase down extra dependencies, I hate the idea of having excessively large API and the same libraries used included and shaded into everything more...

**Comment by jacklin213** [comment]
Tbh can't we allow both options of either a build script or a shadded jar.

Especially for a rapidly changing project like Sponge, not everyone can keep up dependencies. Correct me if I'm wrong but for the people who haven't used build scripts before they don't have to go out of their way to learn about build scripts before trying to develop their idea into sponge. 

I quite like the idea of what @ConnorStone said
* API jar to the build path
* Profit

If we could support both that would be fantastic

**Comment by ST-DDT** [comment]
The none shaded jar should be the recommended one.
But there should also be a shaded one.

The main issue with shaded jars are tutorials on YT or other platforms. Beginners will follow their lead. Videos outrank every textual documentation we could provide on our page. So we must catch all those tutorial guided people and show them  a video and/or documentation on the download page. So the user starts thinking its easier using a build tool. They may not listen to it the first time but after a while they will at least give it a chance when they downloaded the API for the third time.

Beginners count the steps and how complex they sound not the real work.
And especially for those beginners who have neber heart of java or programming at all a shaded jar is a must have.

* Download IDE
* Download Shaded-API
* Create Java-Project
* Add API to class path
* Type the **1** class code provided in the video
* Feel happy that there are no unexplainable red lines
* Export jar to the mods folder
* Test plugin and be happy and proud of yourself

No XML or similar stuff because it looks like dark magic and you hardly take the time to explain it in a tutorial. Bukkits plugin.yml is an exception to this rule as it is quite simple. (No brakets and only 4 lines you can write on your own)

**Comment by ElgarL** [comment]
I have never liked the move to pushing build tools (or any tool) on everyone for every project. If you are creating something huge with a large team contributing then it has its advantages, and often for the more advanced programmers its a must. However, for the person creating plugins in their spare time its an unnecessary extra step and definitely should not be a requirement.

In the corporate environment build tools are a requirement. But here? After all this is a 'fun hobby' and not a 'business'. We are all doing this to enjoy ourselves, not to make money or force some corporate set of rules on people just because its what we were taught when we entered the work force.

In my opinion 'encouraging' people to always use a build tool is a bad practice. There has always been a great idiom of 'the right tool for the job', and as I said earlier KISS (Keep It Simple, Stupid).
Never over complicate things if you don't need to, as the best way to learn is from the simplest of beginnings.
I believe whole heartedly that to learn to program you should do it from the very ground up. Starting with the simplest things up until you are using a full IDE, and any other tools you want to use to make life easier. But you should not be using those tools if you don't know how it is all working under the hood.
Too many programmers these days are sloppy and lazy because they have never known working with severe memory constraints, and many have no idea what 'machine code' or 'assembly' is. Learning these things makes a person a more efficient programmer but are so often skipped as people are thrown in at the deep end, handed a dozen tools and told "it doesn't matter what they do, just use them and it will make your life easier".

Start with the most basic of tools and build up from there. If your project grows to a state of requiring a build tool then it should be adopted as an option, not forced upon every person who wants to dabble with the inner workings of a computer.

**Comment by jacklin213** [comment]
My question is (if this hasnt already been answered) why @Mumfrey doesn't support the use of shaded API (my apologies if this isn't the case).

**Comment by kukelekuuk00** [comment]
Although I don't quite understand why anyone wouldn't use ant+ivy or Maven considering eclipse and intellij both have these things built in or easily available through plugins. I can see the merit in having an official "not recommended" download with a shaded jar.

**Comment by ElgarL** [comment]
I don't see the reason for 'not recommended'. API's have been complete/shaded for ever in the past. This new policy seems recent.

**Comment by Mumfrey** [comment]
> The none shaded jar should be the recommended one.
> But there should also be a shaded one.
> 
> The main issue with shaded jars are tutorials on YT or other platforms. Beginners will follow their lead. Videos outrank every textual documentation we could provide on our page. So we must catch all those tutorial guided people and show them a video and/or documentation on the download page. So the user starts thinking its easier using a build tool. They may not listen to it the first time but after a while they will at least give it a chance when they downloaded the API for the third time.

This doesn't really hold any water as an argument. New coders and people following tutorials for Sponge dev will use the SDK, existing users of Maven, Gradle, Ivy, SBT or any other system supporting dependency management will just consume the API via the published artefact. Are you somehow suggesting that people are going to follow tutorials for Bukkit dev and assume it applies to Sponge? If they do that then the environment is going to be the least of their worries, the entire setup is different for sponge (and in fact is much simpler).

> Beginners count the steps and how complex they sound not the real work.
> And especially for those beginners who have neber heart of java or programming at all a shaded jar is a must have.
> 
>     Download IDE
>     Download Shaded-API
>     Create Java-Project
>     Add API to class path
>     Type the 1 class code provided in the video
>     Feel happy that there are no unexplainable red lines
>     Export jar to the mods folder
>     Test plugin and be happy and proud of yourself
> 
> No XML or similar stuff because it looks like dark magic and you hardly take the time to explain it in a tutorial. Bukkits plugin.yml is an exception to this rule as it is quite simple. (No brakets and only 4 lines you can write on your own)

Again, this doesn't make sense when you offer it up against reality. The *reality* is that people will use the SDK (Bukkit didn't **have** an SDK) which gives them a build tool out-of-the-box and they won't even *realise* they're using a build tool! All they will know is that they don't have to do any work to set up their plugin (like they were forced to do with bukkit). With our system, the "getting started" steps resolve to:

* Download IDE
* Download SDK
* Run the SDK setup task
* Import the generated project direct into their SDK
* Type out their class files and modify the already-provided build script (if they want to)
* Feel happy there are no unexplainable red lines
* **Use the SDK to automagically test their code in a pre-prepared Sponge server environment that "just works"**
* **Use the SDK to export the jar for their project**

As an optional extra they can manage the gradle setup within their IDE, but they don't have to do this, and building, testing and setup are all built in to the SDK with **no** aditional steps required at all.

> I have never liked the move to pushing build tools (or any tool) on everyone for every project. If you are creating something huge with a large team contributing then it has its advantages, and often for the more advanced programmers its a must. However, for the person creating plugins in their spare time its an unnecessary extra step and definitely should not be a requirement.

This is (I think) the fundamental disagreement here. You see build tools as an *extra* step, whereas in general this is just not the case, build tools simply *replace* steps which would otherwise be manual. They don't *add* work, they just change the nature of work you would have to do anyway, and automate it ever after.

See with a manual process you:

* Download dependencies manually
* Add them to your project manually
* Put them in your artefact manually

With a build system you:

* Specify your dependencies
* All other steps happen automatically

Whilst you don't *have* to use a build system from step one, for beginners it makes a **lot** of sense because all the complexity of having to understand and manage dependencies, and put them in jars and suchlike is squirreled away in a nice file in the background which takes care of it all, and all they need to know is that they just run "setup" and "build" steps from the SDK. Having the build system in place **also** nicely prepares them for when their project gets larger in the future, but that's a *bonus* which goes on top of the benefits they get right out of the gate by using the SDK from day one.

Suggesting that forcing new users to do everything the long-winded manual way (which doesn't give them the extra benefits of a built-in testing environment as well for example) is simpler than them just running a single command which forms part of the SDK just doesn't really add up.

Now, please note. **I am not saying your opinion is wrong.** It's your opinion, you're entitled to it, and I agree that there's nothing fundamentally wrong with wanting to do things your own way, as I mentioned above I'm exactly the same and still use Ant for pretty much everything. *However*, that has to stop short of actually recommending new users should also adopt your MO, when there are simpler, more effective ways of working with the API which are available already and work extremely well.

It has to be clear to new users that there's a straightforward way of getting up and running with the API (using the SDK) and all they need to do is "X -> Y -> Z" and they're good to go. We of course support people who are already using Maven, Ivy, Gradle or SBT for their plugins and there's obviously a super-simple way of integrating into those workflows as well. However, if you go "off the beaten track" and want to go down your own route, that's entirely your own business and you're totally entitled to support with the API, etc. etc. and we're all here to lend a hand with that. But then to turn around and say "my way is better and you should change everything to suit me" just *doesn't make sense.*

However, as I already offered. If you want support for your workflow, I am *personally* happy to accomodate you and provide an artefact which you can consume, and as I said that's an open offer and I'm happy to accomodate that at any time. But I really, really want you to be able to understand why it can't be an officially-supported thing. It's just a bad idea for Sponge as an entity to support this kind of workflow.

> In the corporate environment build tools are a requirement. But here? After all this is a 'fun hobby' and not a 'business'. We are all doing this to enjoy ourselves, not to make money or force some corporate set of rules on people just because its what we were taught when we entered the work force.

Nobody's forcing anyone to do anything. As I mentioned above, providing an SDK makes sense (it provides a unified and well-understood environment to provide support for end users for one, and hides away the complexity of dependencies and building jars for another) and doesn't "lock in" anyone if they don't want to use it, they're free to use a build tool they're more comfortable with (like Ant, Maven or SBT) if they want to. But from an official perspective, having a "do it this way" just makes a *lot* of sense because supporting users is tricky at the best of times, and if they're using some unknown setup it's going to be much harder to provide that support, at least effectively.

> In my opinion 'encouraging' people to always use a build tool is a bad practice.

That's your opinion, and it's the part of your argument which I'm directly contesting. I think - on the contrary - encouraging people to use a build tool is the only sane thing. In *exactly* the same way that encouraging people to use an IDE is a sane thing. Do you see where I'm coming from?

>There has always been a great idiom of 'the right tool for the job', and as I said earlier KISS (Keep It Simple, Stupid).

Completely agree. As a software developer by profession I live my life by the KISS principle. And the SDK with "setup this", "build that" **is where that simplicity lies** in this scenario.

> Never over complicate things if you don't need to, as the best way to learn is from the simplest of beginnings.

Exactly!

> I believe whole heartedly that to learn to program you should do it from the very ground up. Starting with the simplest things up until you are using a full IDE, and any other tools you want to use to make life easier. But you should not be using those tools if you don't know how it is all working under the hood.

This just doesn't work as a philosophy. Remember when you started java and didn't really understand why you had to type `public static void main(String[] args)` but that meant you could learn the syntax and later figure out why you had to type that stuff? The solution we have means you can "dive right in" to learning to write plugins and write code (and test it) by just going through some simple motions which handle the setup automatically, then once you're ready to tackle the additional steps you can do so at your own pace, the point is you're not forced at any stage to understand something which isn't necessary for you to understand at that particular moment.

Again, if that's your opinion I can see where you're coming from. But you have to understand that there are *alternate* opinions which disagree, and will provide a better experience for new and intermediate users who are just following tutorials to get started.

> Too many programmers these days are sloppy and lazy because they have never known working with severe memory constraints, and many have no idea what 'machine code' or 'assembly' is. 

This is getting off topic. I quite agree, but it's not relevant for the discussion. And also the implication that you have to learn the nuts and bolts *first* is potentially a disputable point.

> Learning these things makes a person a more efficient programmer but are so often skipped as people are thrown in at the deep end, handed a dozen tools and told "it doesn't matter what they do, just use them and it will make your life easier".

Yes, this is the nature of the beast nowadays. Believe me, I'm a miserable old git but even I acknowledge that the game has changed. You can live in denial of that fact as much as you like but reality won't change to suit you I'm afraid.

> Start with the most basic of tools and build up from there.

That's what we do. We just define "basic" as "not confusing", rather than "bare bones".

> If your project grows to a state of requiring a build tool then it should be adopted as an option, not forced upon every person who wants to dabble with the inner workings of a computer.

Sorry but I disagree, for all the reasons outlined above.

**Comment by ElgarL** [comment]
You see your steps of...

    Download IDE (possibly install Maven, Gradle or a plugin for ivy etc)
    Download SDK
    Run the SDK setup task
    Import the generated project direct into their SDK
    Type out their class files and modify the already-provided build script (if they want to)
    Feel happy there are no unexplainable red lines
    Use the SDK to automagically test their code in a pre-prepared Sponge server environment that "just works"
    Use the SDK to export the jar for their project

as simpler than...

    Download IDE
    Add API jar to the build path
    Type out their class files
    Feel happy there are no unexplainable red lines
    Export the jar.

The second is preferable, but only possible with a shaded jar. Its quicker, simpler and more straight forward. The SDK requires Gradle and a lot of extra steps. Its the complexity of the SDK steps that turned a lot of people off of modding on Forge.
I'm arguing for an officially provided and supported shaded jar because (although your offer is kind) relying on the good will of an individual dev to provide a shaded jar in perpetuity would be silly.

**Comment by ST-DDT** [comment]
@Mumfrey I will be happy if everyone releases (open source) software using standardized build tools. And i myself use them in most cases. (In offline environments this is not possible all the time)

My YT-Point is that there will be tutorials that will not use the SDK however using it might be desirable.
We must assume the worst (tutorial) possible and then guide the missleaded souls to either using the SDK or build tools. The issue are not the tutorials provided by the sponge team. The issue are those tutorial publishers that don't like our SDK or any build tools and thus don't use them in their tutorials.

And if we don't provide the shaded jar by ourselfes, somebody else will do that and the downloaders will blame us if dependency xy is missing in that jar or there are some malicious code changes in there or the jar simply does not work.
And now combine this with the quality of bug reports (most) beginners submit.
We have to ask for a stacktrace, ask where have they downloaded the version, the code that actually fails (which might work with a proper API download) ....

**Pros providing a shaded jar or a zip containing all dependency jars):**
* Does not cost us anything (work and time) except some storage and bandwidth
* Allows us to measure how many people follow that path at a time
* Provides us with the ability to show them a hint/tutorial how to do things easier
* Show the downloaders the right and trustworth portal to ask their questions and download other jars.
* Generate some minor popularity increase in googles search algorithm for our page as we get linked from more pages, that would otherwise link to some third party page providing a shaded jar

**Cons**
- Some extra storage and bandwidth usage
- Allows starters to do things wrong more easily
  - But we can keep the control and try to establish countermeasures

**Summary:** 
- Pro Shaded, because we can
- Pro Build-Tools, because it is easier
  - And this is the reason why the shaded.jar shouldn't be the recommended way

PS: Don't call the shaded jar `SpongeAPI-all.jar` or `SpongeAPI-full.jar` as it is very misleading.

@ElgarL I guess most people got/understand your point/thoughts already. 
But please agree with us that we disagree with you on this topic.
You and me aren't members of the official staff so in the end it is not our decision.
The discussion starts circling. I think we should both sit down and wait for others to bring in new points. Otherwise the Collaborators will stop listening to our arguments.

**Comment by MattBDev** [comment]
We still haven't heard the opinion from the OP on this issue.

**Comment by kashike** [comment]
@MattBDev I didn't see a need to reply and re-state what has been said, but:

As @Mumfrey said:

> We add a task in the API build.gradle file to emit a fully shaded jar, which would allow you to build it and host it on your server, which is what I assume you want to do ultimately.
We do that, but I set up my jenkins (ci.liteloader.com) to also build that shaded jar and give you a place to download it from.
If either of those are helpful (or both), then I can set that up. It keeps the fully shaded jar off the official Sponge repositories but gives you somewhere to get what you need.

and @ElgarL said:
> Its a nice offer and I believe a shaded jar is the correct way to go, but I feel it needs to be an officially available option. I think it should be the default option over a build system favoured one.

I'd like to see an "officially-provided" shadowed JAR in addition to the current JARs.

**Comment by Zidane** [comment]
@ElgarL @MattBDev @kashike 

(copied from the other controversial issue regarding config formats)

Ultimately, I end up making the call as the lead the API and Vanilla implementation. Do not take what one or more members say as an official statement on this matter.

I'm closing this until I make my statement on Sponge's stance.

**Comment by kashike** [comment]
@Zidane Is there a decision on this?

**Comment by Zidane** [comment]
@kashike 

After sitting on this for a while and hearing all sides, I'm ready to make my decision.

I've decided to make the call to provided a shadowed jar. This comes with a few conditions

* This will be provided as an option to download off of our downloads page and no-where else
* We will provide ZERO support for a setup that involves just having the API-combined jar. If you are having issues writing a plugin by using the  combined jar in your favorite IDE/whatever, you are on your own
* Our official stance remains with build systems (gradle and maven). This won't be changing.

This covers both parties; we keep our responsibility to a minimum by providing the jar and that's it. 

@progwml6 as soon as your schedule allows, please add the entry to our downloads site (figure out how you want to show it)

@ElgarL Once the above is in place, I'll let you know.

Just in-case I wasn't clear above, this decision is FINAL.
